Add documentation pages on 2D and 3D antialiasing (#6267)
godotengine/godot-docs: tutorials/3d/3d_antialiasing.rst

In the example below, you can notice how
edges have a blocky appearance. The vegetation is also flickering in and out,
and thin lines on top of the box have almost disappeared:

.. figure:: img/antialiasing_none_scaled.png
   :alt: Image is scaled by 2× with nearest-neighbor filtering to make aliasing more noticeable.
   :align: center

   Image is scaled by 2× with nearest-neighbor filtering to make aliasing more noticeable.

To combat this, various antialiasing techniques can be used in Godot. These are
detailed below.

Multisample antialiasing (MSAA)
-------------------------------

This technique is the "historical" way of dealing with aliasing. MSAA is very
effective on geometry edges (especially at higher levels). MSAA does not
introduce any blurriness whatsoever.

MSAA in available in 3 levels: 2×, 4×, 8×. Higher levels are more effective at
antialiasing edges, but are significantly more demanding. In games with modern
visuals, sticking to 2× or 4× MSAA is highly recommended as 8× MSAA is usually
too demanding.

The downside of MSAA is that it only operates on edges. This is because MSAA
increases the number of *coverage* samples, but it still runs shaders for each

> pixel only once. Therefore, MSAA does not reduce transparency aliasing for

Reviewer: The term "coverage samples" seems important here, but isn't defined.

Author: I've added a definition of coverage samples.

    pixels.

    If you are using a high window size such as 4K, you may find that increasing
    the resolution scale past a certain value will cause a heavy slowdown (or
    even a crash) due to running out of VRAM.

Screen-space roughness limiter
------------------------------

*This is only available in the Clustered Forward and Forward Mobile backends,
not the Compatibility backend.*

This is not an edge antialiasing method, but it is a way of reducing specular
aliasing in 3D.

The screen-space roughness limiter works best on detailed geometry. While it has
an effect on roughness map rendering itself, its impact is limited there.

The screen-space roughness limiter is enabled by default; it doesn't require
any manual setup. It has a small performance impact, so consider disabling it
if your project isn't affected by specular aliasing much.

Texture roughness limiter on import
-----------------------------------

Like the screen-space roughness limiter, this is not an edge antialiasing
method, but it is a way of reducing specular aliasing in 3D.

Roughness limiting on import works by specifying a normal map to use as a guide

> for limiting roughness. This is done in the by selecting the roughness map in the

Reviewer: ```suggestion
for limiting roughness. This is done by selecting the roughness map in the
```

FileSystem dock, then going to the Import dock and setting **Roughness > Mode**
to the color channel the roughness map is stored in (typically **Green**), then
setting the path to the material's normal map. Remember to click **Reimport**
at the bottom of the Import dock after setting the path to the normal map.

Since this processing occurs purely on import, it has no performance cost
whatsoever. However, its visual impact is limited. Limiting roughness on import

> only helps reduce specular aliasing within textures, not the one present on

Reviewer: ```suggestion
only helps reduce specular aliasing within textures, not the aliasing that occurs on
```

geometry edges on detailed meshes.

Which antialiasing technique should I use?
------------------------------------------

**There is no "one size fits all" antialiasing technique.** Since antialiasing is
often demanding on the GPU or can introduce unwanted blurriness, you'll want to
add a setting to allow players to disable antialiasing.

For projects with a photorealistic art direction, TAA is generally the most

> suited option. While it introduces ghosting artifacts, there is no other

Reviewer: ```suggestion
suitable option. While it introduces ghosting artifacts, there is no other
```

> technique that combats specular aliasing as well as TAA does. The screen-space
> roughness limiter helps a little, but is far less effective against specular
> aliasing after all.

Reviewer: ```suggestion
technique that combats specular aliasing as well as TAA does. The screen-space
roughness limiter helps a little, but is far less effective against specular
aliasing overall.
```
Does this preserve the intended meaning?

For projects with a low amount of reflective surfaces (such as a cartoon
artstyle), MSAA can work well. MSAA is also a good option if avoiding blurriness
and temporal artifacts is important, such as in competitive games.

When targeting low-end platforms such as mobile or integrated graphics, FXAA is
usually the only viable option. 2× MSAA may be usable in some circumstances,
but higher MSAA levels are unlikely to run smoothly on mobile GPUs.

Godot allows using multiple antialiasing techniques at the same time. This is

> usually not needed, but it can provide better visuals on high-end GPUs or for

Reviewer: ```suggestion
usually unnecessary, but it can provide better visuals on high-end GPUs or for
```

.. _doc_3d_antialiasing:

3D antialiasing
===============

.. Images on this page were generated using the project below
.. (except for `antialiasing_none_scaled.png`):
.. https://github.com/Calinou/godot-antialiasing-comparison

.. seealso::

    Godot also supports antialiasing in 2D rendering. This is covered on the
    :ref:`doc_2d_antialiasing` page.

Introduction
------------

Due to their limited resolution, scenes rendered in 3D can exhibit aliasing

> artifacts. These artifacts manifest in the form of a "staircase" effect on
> geometry edges, but also with reflective surfaces flickering in and out
> (specular aliasing).

Reviewer 2: ```suggestion
artifacts. These artifacts commonly manifest as a "staircase" effect
on surface edges (edge aliasing) and as  flickering and/or
sparkles on reflective surfaces (specular aliasing).
```

In the example below, you can notice how
edges have a blocky appearance. The vegetation is also flickering in and out,
and thin lines on top of the box have almost disappeared:

.. figure:: img/antialiasing_none_scaled.png
   :alt: Image is scaled by 2× with nearest-neighbor filtering to make aliasing more noticeable.
   :align: center

   Image is scaled by 2× with nearest-neighbor filtering to make aliasing more noticeable.

To combat this, various antialiasing techniques can be used in Godot. These are
detailed below.

Multisample antialiasing (MSAA)
-------------------------------

This technique is the "historical" way of dealing with aliasing. MSAA is very
effective on geometry edges (especially at higher levels). MSAA does not
introduce any blurriness whatsoever.

> MSAA in available in 3 levels: 2×, 4×, 8×. Higher levels are more effective at

Reviewer 2: ```suggestion
MSAA is available in 3 levels: 2×, 4×, 8×. Higher levels are more effective at
```

StandardMaterial3D or ORMMaterial3D properties. This only has an effect when
MSAA is used (with any level). Alpha to coverage has a moderate performance
cost, but it's very effective at reducing aliasing on transparent materials
without introducing any blurriness.

MSAA can be enabled in the Project Settings by changing the value of the
**Rendering > Anti Aliasing > Quality > MSAA 3D** setting. It's important to change
the value of the **MSAA 3D** setting and not **MSAA 2D**, as these are entirely
separate settings.

Comparison between no antialiasing (left) and various MSAA levels (right).
Note that alpha antialiasing is not used here:

.. image:: img/antialiasing_msaa_2x.png

.. image:: img/antialiasing_msaa_4x.png

.. image:: img/antialiasing_msaa_8x.png

Temporal antialiasing (TAA)
---------------------------

*This is only available in the Clustered Forward backend, not the Forward Mobile
or Compatibility backends.*

Temporal antialiasing works by *converging* the result of previously rendered
frames into a single, high-quality frame. This is a continuous process that
works by jittering the position of all vertices in the scene every frame. This

> jittering is done to allow for antialiasing to work when the camera isn't
> moving, and is generally unnoticeable.

Reviewer 2: ```suggestion
jittering is done to capture sub-pixel detail and should be unnoticeable except
in extreme situations.
```

geometry edges, but also with reflective surfaces flickering in and out
(specular aliasing).

In the example below, you can notice how
edges have a blocky appearance. The vegetation is also flickering in and out,
and thin lines on top of the box have almost disappeared:

.. figure:: img/antialiasing_none_scaled.png
   :alt: Image is scaled by 2× with nearest-neighbor filtering to make aliasing more noticeable.
   :align: center

   Image is scaled by 2× with nearest-neighbor filtering to make aliasing more noticeable.

To combat this, various antialiasing techniques can be used in Godot. These are
detailed below.

Multisample antialiasing (MSAA)
-------------------------------

This technique is the "historical" way of dealing with aliasing. MSAA is very
effective on geometry edges (especially at higher levels). MSAA does not
introduce any blurriness whatsoever.

MSAA in available in 3 levels: 2×, 4×, 8×. Higher levels are more effective at
antialiasing edges, but are significantly more demanding. In games with modern
visuals, sticking to 2× or 4× MSAA is highly recommended as 8× MSAA is usually
too demanding.

> The downside of MSAA is that it only operates on edges. This is because MSAA
> increases the number of *coverage* samples, but not the number of *color*
> samples. Using multiple coverage samples allows the coverage value to be an
> intermediate value between 0 and 1, rather than being limited to 0.0 *or* 1.0.
> The coverage value is used to determine the contribution of the pixel to the
> scene, with 0.0 meaning "no contribution at all" and 1.0 meaning "fully opaque
> pixel".
> 
> However, since the number of color samples did not increase, fragment shaders
> are still run for each pixel only once. Therefore, MSAA does not reduce
> transparency aliasing for materials using the **Alpha Scissor** transparency
> mode (1-bit transparency). MSAA is also ineffective on specular aliasing.

Reviewer: ```suggestion
MSAA has the downside of being ineffective on specular aliasing. It also only operates
on edges - it does not reduce transparency aliasing for materials using the **Alpha Scissor**
transparency mode (1-bit transparency).
```

While I like a good delve into technical details, would it be better to keep the section focused only on outlining the downsides of MSAA?

I also tried to move specular aliasing first so that the alpha scissoring problem flows into the next paragraph on mitigating it, but I'm not very happy with my wording.

Author: I can remove the description of coverage samples if needed – I'm not sure if it's actually required to understand what's going on.

Author: Copying the part I removed for future reference:

```rst
Using multiple coverage samples allows the coverage value to be an
intermediate value between 0 and 1, rather than being limited to 0.0 *or* 1.0.
The coverage value is used to determine the contribution of the pixel to the
scene, with 0.0 meaning "no contribution at all" and 1.0 meaning "fully opaque
pixel".
```
